indico/web/client/js/custom_elements/ind_share_widget.js

```diff
@@ -0,0 +1,362 @@
+    },
+  ];
+
+  return (
+    <Grid columns={2} stretched>
+      {calendars.map(calendar => (
+        <GridColumn key={calendar.name} styleName="share-button-column">
+          <Button
+            as="a"
+            href={calendar.url + calendar.params}
+            target="_blank"
+            basic
+            color={calendar.color}
+            styleName="share-button"
+          >
+            <img src={calendar.logo} alt={Translate.string(calendar.name)} />
+            <Translate as="span">{calendar.name}</Translate>
+          </Button>
+        </GridColumn>
+      ))}
+    </Grid>
+  );
+}
+
+CalendarButtons.propTypes = {
+  googleCalParams: PropTypes.string.isRequired,
+  outlookCalParams: PropTypes.string.isRequired,
+};
+
+function SocialButton({name, logo, url, color, render, button, buttonProps}) {
+  const buttonElement = button ? (
+    <Button basic color={color} styleName="share-button" {...buttonProps}>
+      <img src={logo} alt={Translate.string(name)} />
+      <span>{name}</span>
+    </Button>
+  ) : (
+    <Button as="a" href={url} target="_blank" basic color={color} styleName="share-button">
+      <img src={logo} alt={Translate.string(name)} />
+      <span>{name}</span>
+    </Button>
+  );
+  return (
+    <GridColumn styleName="share-button-column">
+      {render ? render(buttonElement) : buttonElement}
+    </GridColumn>
+  );
+}
+
+SocialButton.propTypes = {
+  name: PropTypes.string.isRequired,
+  logo: PropTypes.string.isRequired,
+  url: PropTypes.string,
+  color: PropTypes.string.isRequired,
+  render: PropTypes.func.isRequired,
+  button: PropTypes.bool,
+  buttonProps: PropTypes.object,
+};
+
+SocialButton.defaultProps = {
+  url: null,
+  button: false,
+  buttonProps: {},
+};
+
+function SetupMastodonServer({setMastodonOpen, button, urlParams}) {
+  const [error, setError] = useState(false);
+  if (Indico.User.mastodonServerURL) {
+    return button;
+  }
+
+  return (
+    <Popup
+      wide
+      pinned
+      position="top right"
+      onOpen={() => setMastodonOpen(true)}
+      onClose={() => {
+        setError(false);
+        setMastodonOpen(false);
+      }}
+      trigger={button}
+      on="click"
+    >
+      <Popup.Content>
+        <Message icon info styleName="empty-mastodon-server-notice">
+          <Icon name="star" />
+          <Translate>
+            You have not added a preferred Mastodon server on your profile yet. Please add one
+            below.
+          </Translate>
+        </Message>
+        <FinalForm
+          onSubmit={async ({mastodonServerURL}) => {
+            try {
+              await indicoAxios.post(userPreferencesMastodonServer(), {
+                server_url: mastodonServerURL,
+              });
+              setError(null);
+            } catch (e) {
+              if (_.get(e, 'response.status') !== 418) {
+                return handleAxiosError(e);
+              }
+              setError(e.response.data.message);
+              return;
```

Comment:
Why all this custom logic? The standard `return handleSubmitError(e);` should work - final-form will then set the error on the field, and you can access that inside the field's `render` callback.

Reply:
BTW, see if you can't use `FinalField` (or rather `FinalInput`) for the component since it already handles all the error display logic etc. Not sure how well it works inside the popup though since it's not displayed as a "standard" form...